### PR TITLE
OpenEdu SmartExcel

### PR DESCRIPTION
**Pull request by pierrealixt**

### CODE REVIEW

**Review by Christiaanvdm — APPROVED** [review]
LGTM, just make sure you fix the Travis pep8 and test error!

File: django_project/ford3/models/provider.py [hunk 1]
```diff
@@ -267,28 +267,61 @@ def get_location_as_dict(self):
...
-                    setattr(model, property_name, property_value)
-                    model.save()
+                if key != 'qualification__id':
+                    if key.find('--') > 0:
+                        original_key = key
+                        key = key[:key.find('--')]
+                        key_index = key[key.find('--')+2:]
+                    else:
+                        original_key = key
+                        key_index = 0
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py]: original_key, key_index = determine_stuff(key)

File: django_project/ford3/models/provider.py [hunk 2]
```diff
@@ -267,28 +267,61 @@ def get_location_as_dict(self):
...
+                    if key == 'qualification_entrance_requirement_subject__subject':
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py]: pass the code in a function

File: django_project/ford3/models/provider.py [hunk 3]
```diff
@@ -267,28 +267,61 @@ def get_location_as_dict(self):
...
-                    model_name = key[:key.find('__')]
-                    try:
-                        model = models[model_name]
-                    except KeyError as e:
-                        errors[i][key] = str(e)
-                    property_name = key[key.find('__')+2:]
-                    property_value = obj[key]
-                    setattr(model, property_name, property_value)
-                    model.save()
+                if key != 'qualification__id':
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py]: inverse the condition and `continue`

File: django_project/ford3/models/provider.py [hunk 4]
```diff
@@ -267,28 +267,61 @@ def get_location_as_dict(self):
...
             models['saqa_qualification'] = current_qualification.saqa_qualification
             models['qualification'] = current_qualification
             models['requirement'] = current_qualification.requirement
-            if models['requirement']:
+
+            if not models['requirement']:
                 requirement = (apps.get_model('ford3', 'Requirement'))()
                 requirement.qualification_id = current_qualification.id
                 requirement.save()
                 models['requirement'] = requirement
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py]: you could have a method to return a `models` dictionary.

File: django_project/ford3/models/provider.py [hunk 5]
```diff
@@ -267,28 +267,61 @@ def get_location_as_dict(self):
...
+                                    models['qualification_entrance_requirement_subject'] = (
+                                        current_qualification.qualificationentrancerequirementsubject_set.all()[key_index])
+
+                                except IndexError:
+                                    qualification_entrance_requirement_subject = (
+                                        qualification_entrance_requirement_subject_model.objects.create(
+                                            qualification=current_qualification, subject=property_value))
+                                    models['qualification_entrance_requirement_subject'] = (
+                                        qualification_entrance_requirement_subject)
+                    else:
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py]: add comment of what it does

File: django_project/ford3/models/provider.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 from django.contrib.gis.db.models import PointField
 from django.contrib.gis.geos import Point, GEOSGeometry
 from django.apps import apps
+# from django.core.exceptions import ObjectDoesNotExist
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py:7]: Remove if not used.

File: django_project/ford3/views/views.py [hunk 7]
```diff
@@ -44,5 +44,55 @@ def delete_qualification(request, provider_id, campus_id, qualification_id):
...
+
+def diff(key, obj, new_value):
+    if getattr(obj, key) != new_value:
+        setattr(obj, key, new_value)
+        obj.save()
+        return True
+    return False
+
+
+def import_qualification(request, provider_id):
```
Comment (pierrealixt) [inline django_project/ford3/views/views.py:53]: is it still used?

File: django_project/ford3/models/provider.py [hunk 8]
```diff
@@ -262,3 +265,149 @@ def get_location_as_dict(self):
                 'location_value_x': 0,
                 'location_value_y': 0})
         return result
+
+    def import_excel_data(self, data):
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py]: how do we call this method?

Comment (pierrealixt) [reply django_project/ford3/models/provider.py]: I feel this method (and the others) should be somewhere else than in the Provider model.

File: django_project/ford3/models/provider.py [hunk 9]
```diff
@@ -262,3 +265,149 @@ def get_location_as_dict(self):
...
+                    model_name = key[:key.find('__')]
+                    try:
+                        model = models[model_name]
+                    except KeyError as e:
+                        errors[idx][key] = str(e)
+                    property_name = key[key.find('__')+2:]
+                    property_value = obj[key]
+                    if property_value and model:
+                        setattr(model, property_name, property_value)
+                        model.save()
```
Comment (pierrealixt) [inline django_project/ford3/models/provider.py]: what about the `errors` dict?
What does this method return?

**Pull request merged**